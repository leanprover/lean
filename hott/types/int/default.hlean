/-
Copyright (c) 2014 Floris van Doorn. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Floris van Doorn
-/

import .basic .hott
