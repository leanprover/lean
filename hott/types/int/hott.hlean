/-
Copyright (c) 2015 Floris van Doorn. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Floris van Doorn

Theorems about the integers specific to HoTT
-/

import .basic types.eq arity algebra.bundled
open core eq is_equiv equiv algebra is_trunc
open nat (hiding pred)

namespace int

  section
  open algebra
  definition group_integers : Group :=
  Group.mk ℤ (group_of_add_group _)
  end

  definition is_equiv_succ [instance] : is_equiv succ :=
  adjointify succ pred (λa, !add_sub_cancel) (λa, !sub_add_cancel)
  definition equiv_succ : ℤ ≃ ℤ := equiv.mk succ _

  definition is_equiv_neg [instance] : is_equiv (neg : ℤ → ℤ) :=
  adjointify neg neg (λx, !neg_neg) (λa, !neg_neg)
  definition equiv_neg : ℤ ≃ ℤ := equiv.mk neg _

  definition iterate {A : Type} (f : A ≃ A) (a : ℤ) : A ≃ A :=
  rec_nat_on a erfl
               (λb g, f ⬝e g)
               (λb g, g ⬝e f⁻¹ᵉ)

  -- definition iterate_trans {A : Type} (f : A ≃ A) (a : ℤ)
  --   : iterate f a ⬝e f = iterate f (a + 1) :=
  -- sorry

  -- definition trans_iterate {A : Type} (f : A ≃ A) (a : ℤ)
  --   : f ⬝e iterate f a = iterate f (a + 1) :=
  -- sorry

  -- definition iterate_trans_symm {A : Type} (f : A ≃ A) (a : ℤ)
  --   : iterate f a ⬝e f⁻¹e = iterate f (a - 1) :=
  -- sorry

  -- definition symm_trans_iterate {A : Type} (f : A ≃ A) (a : ℤ)
  --   : f⁻¹e ⬝e iterate f a = iterate f (a - 1) :=
  -- sorry

  -- definition iterate_neg {A : Type} (f : A ≃ A) (a : ℤ)
  --   : iterate f (-a) = (iterate f a)⁻¹e :=
  -- rec_nat_on a idp
  --   (λn p, calc
  --     iterate f (-succ n) = iterate f (-n) ⬝e f⁻¹e : rec_nat_on_neg
  --       ... = (iterate f n)⁻¹e ⬝e f⁻¹e : by rewrite p
  --       ... = (f ⬝e iterate f n)⁻¹e : sorry
  --       ... = (iterate f (succ n))⁻¹e : idp)
  --   sorry

  -- definition iterate_add {A : Type} (f : A ≃ A) (a b : ℤ)
  --   : iterate f (a + b) = equiv.trans (iterate f a) (iterate f b) :=
  -- sorry

  -- definition iterate_sub {A : Type} (f : A ≃ A) (a b : ℤ)
  --   : iterate f (a - b) = equiv.trans (iterate f a) (equiv.symm (iterate f b)) :=
  -- sorry

  -- definition iterate_mul {A : Type} (f : A ≃ A) (a b : ℤ)
  --   : iterate f (a * b) = iterate (iterate f a) b :=
  -- sorry

end int open int



namespace eq
  variables {A : Type} {a : A} (p : a = a) (b c : ℤ) (n : ℕ)
  definition power : a = a :=
  rec_nat_on b idp
               (λc q, q ⬝ p)
               (λc q, q ⬝ p⁻¹)
  --iterate (equiv_eq_closed_right p a) b idp

  -- definition power_neg_succ (n : ℕ) : power p (-succ n) = power p (-n) ⬝ p⁻¹ :=
  -- !rec_nat_on_neg

  -- local attribute nat.add int.add int.of_num nat.of_num int.succ [constructor]

  definition power_con : power p b ⬝ p = power p (succ b) :=
  rec_nat_on b
    idp
    (λn IH, idp)
    (λn IH, calc
      power p (-succ n) ⬝ p
            = (power p (-int.of_nat n) ⬝ p⁻¹) ⬝ p : by rewrite [↑power,-rec_nat_on_neg]
        ... = power p (-int.of_nat n) : inv_con_cancel_right
        ... = power p (succ (-succ n)) : by rewrite -succ_neg_succ)

  definition power_con_inv : power p b ⬝ p⁻¹ = power p (pred b) :=
  rec_nat_on b
    idp
    (λn IH, calc
      power p (succ n) ⬝ p⁻¹ = power p n : by apply con_inv_cancel_right
        ... = power p (pred (succ n))   : by rewrite pred_nat_succ)
    (λn IH, calc
      power p (-int.of_nat (succ n)) ⬝ p⁻¹
            = power p (-int.of_nat (succ (succ n))) : by rewrite [↑power,-rec_nat_on_neg]
        ... = power p (pred (-succ n)) : by rewrite -neg_succ)

  definition con_power : p ⬝ power p b = power p (succ b) :=
  rec_nat_on b
  ( by rewrite ↑[power];exact !idp_con⁻¹)
  ( λn IH, proof calc
    p ⬝ power p (succ n) = (p ⬝ power p n) ⬝ p : con.assoc p _ p
      ... = power p (succ (succ n)) : by rewrite IH qed)
  ( λn IH, calc
          p ⬝ power p (-int.of_nat (succ n))
                = p ⬝ (power p (-int.of_nat n) ⬝ p⁻¹) : by rewrite [↑power, rec_nat_on_neg]
            ... = (p ⬝ power p (-int.of_nat n)) ⬝ p⁻¹ : con.assoc
            ... = power p (succ (-int.of_nat n)) ⬝ p⁻¹ : by rewrite IH
            ... = power p (pred (succ (-int.of_nat n))) : power_con_inv
            ... = power p (succ (-int.of_nat (succ n))) : by rewrite [succ_neg_nat_succ,int.pred_succ])

  definition inv_con_power : p⁻¹ ⬝ power p b = power p (pred b) :=
  rec_nat_on b
  ( by rewrite ↑[power];exact !idp_con⁻¹)
  (λn IH, calc
    p⁻¹ ⬝ power p (succ n) = p⁻¹ ⬝ power p n ⬝ p : con.assoc
      ... = power p (pred n) ⬝ p : by rewrite IH
      ... = power p (succ (pred n)) : power_con
      ... = power p (pred (succ n)) : by rewrite [succ_pred,-int.pred_succ n])
  ( λn IH, calc
    p⁻¹ ⬝ power p (-int.of_nat (succ n))
          = p⁻¹ ⬝ (power p (-int.of_nat n) ⬝ p⁻¹) : by rewrite [↑power,rec_nat_on_neg]
      ... = (p⁻¹ ⬝ power p (-int.of_nat n)) ⬝ p⁻¹ : con.assoc
      ... = power p (pred (-int.of_nat n)) ⬝ p⁻¹ : by rewrite IH
      ... = power p (-int.of_nat (succ n)) ⬝ p⁻¹ : by rewrite -neg_succ
      ... = power p (-succ (succ n)) : by rewrite [↑power,-rec_nat_on_neg]
      ... = power p (pred (-succ n)) : by rewrite -neg_succ)

  definition power_con_power : Π(b : ℤ), power p b ⬝ power p c = power p (b + c) :=
  rec_nat_on c
    (λb, by rewrite int.add_zero)
    (λn IH b, by rewrite [-con_power,-con.assoc,power_con,IH,↑succ,add.assoc,
                          add.comm (int.of_nat n)])
    (λn IH b, by rewrite [neg_nat_succ,-inv_con_power,-con.assoc,power_con_inv,IH,↑pred,
                          +sub_eq_add_neg,add.assoc,add.comm (-n)])

end eq
