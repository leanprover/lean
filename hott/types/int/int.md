types.int
=========

The integers. Note: most of these files are ported from the standard library. If anything needs to be changed, it is probably a good idea to change it in the standard library and then port the file again (see also [script/port.pl](../../script/port.pl)).

* [basic](basic.hlean) : the integers, with basic operations
* [hott](hott.hlean) : facts about the integers specific to the HoTT library