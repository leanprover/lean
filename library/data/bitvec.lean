/-
Copyright (c) 2015 Joe Hendrix. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Joe Hendrix, Sebastian Ullrich

Basic operations on bitvectors.

This is a work-in-progress, and contains additions to other theories.
-/
import data.tuple

@[reducible] def bitvec (n : ℕ) := tuple bool n

namespace bitvec
open nat
open tuple

local infix `++ₜ`:65 := tuple.append

-- Create a zero bitvector
@[reducible] protected def zero (n : ℕ) : bitvec n := repeat ff n

-- Create a bitvector with the constant one.
@[reducible] protected def one : Π (n : ℕ), bitvec n
| 0        := []
| (succ n) := repeat ff n ++ₜ [tt]

protected def cong {a b : ℕ} (h : a = b) : bitvec a → bitvec b
| ⟨x, p⟩ := ⟨x, h ▸ p⟩

section shift
  variable {n : ℕ}

  def shl (x : bitvec n) (i : ℕ) : bitvec n :=
  let r := dropn i x ++ₜ repeat ff (min n i) in
  have length r = n, begin dsimp, rewrite [nat.sub_add_min_cancel] end,
  bitvec.cong this r

  def fill_shr (x : bitvec n) (i : ℕ) (fill : bool) : bitvec n :=
  let y := repeat fill (min n i) ++ₜ firstn (n-i) x in
  have length y = n, from if h : i ≤ n then
    begin
      dsimp,
      rw [min_eq_right h, min_eq_left (sub_le _ _), -nat.add_sub_assoc h,
        nat.add_sub_cancel_left]
    end
  else
    have h : i ≥ n, from le_of_not_ge h,
    begin
      dsimp,
      rw [min_eq_left h, sub_eq_zero_of_le h, min_eq_left (zero_le _)],
      apply rfl
    end,
  bitvec.cong this y

  -- unsigned shift right
  def ushr (x : bitvec n) (i : ℕ) : bitvec n :=
  fill_shr x i ff

  -- signed shift right
  def sshr : Π {m : ℕ}, bitvec m → ℕ → bitvec m
  | 0        _ _ := []
  | (succ m) x i := head x :: fill_shr (tail x) i (head x)

end shift

section bitwise
  variable {n : ℕ}

  def not : bitvec n → bitvec n := map bnot
  def and : bitvec n → bitvec n → bitvec n := map₂ band
  def or  : bitvec n → bitvec n → bitvec n := map₂ bor
  def xor : bitvec n → bitvec n → bitvec n := map₂ bxor

end bitwise

section arith
  variable {n : ℕ}

  protected def xor3 (x y c : bool) := bxor (bxor x y) c
  protected def carry (x y c : bool) :=
  x && y || x && c || y && c

  def neg (x : bitvec n) : bitvec n :=
  let f := λ y c, (y || c, bxor y c) in
  prod.snd (map_accumr f x ff)

  -- Add with carry (no overflow)
  def adc (x y : bitvec n) (c : bool) : bitvec (n+1) :=
  let f := λ x y c, (bitvec.carry x y c, bitvec.xor3 x y c) in
  let ⟨c, z⟩ := tuple.map_accumr₂ f x y c in
  c :: z

  def add (x y : bitvec n) : bitvec n := tail (adc x y ff)

  protected def borrow (x y b : bool) :=
  bnot x && y || bnot x && b || y && b

  -- Subtract with borrow
  def sbb (x y : bitvec n) (b : bool) : bool × bitvec n :=
  let f := λ x y c, (bitvec.borrow x y c, bitvec.xor3 x y c) in
  tuple.map_accumr₂ f x y b

  def sub (x y : bitvec n) : bitvec n := prod.snd (sbb x y ff)

  instance : has_zero (bitvec n) := ⟨bitvec.zero n⟩
  instance : has_one (bitvec n)  := ⟨bitvec.one n⟩
  instance : has_add (bitvec n)  := ⟨add⟩
  instance : has_sub (bitvec n)  := ⟨sub⟩
  instance : has_neg (bitvec n)  := ⟨neg⟩

  def mul (x y : bitvec n) : bitvec n :=
  let f := λ r b, cond b (r + r + y) (r + r) in
  list.foldl f 0 (to_list x)

  instance : has_mul (bitvec n)  := ⟨mul⟩
end arith

section comparison
  variable {n : ℕ}

  def uborrow (x y : bitvec n) : bool := prod.fst (sbb x y ff)

  def ult (x y : bitvec n) : Prop := uborrow x y
  def ugt (x y : bitvec n) : Prop := ult y x

  def ule (x y : bitvec n) : Prop := ¬ (ult y x)
  def uge (x y : bitvec n) : Prop := ule y x

  def sborrow : Π {n : ℕ}, bitvec n → bitvec n → bool
  | 0        _ _ := ff
  | (succ n) x y :=
     bool.cases_on
       (head x)
       (bool.cases_on (head y) (uborrow (tail x) (tail y)) tt)
       (bool.cases_on (head y) ff (uborrow (tail x) (tail y)))

  def slt (x y : bitvec n) : Prop := sborrow x y
  def sgt (x y : bitvec n) : Prop := slt y x
  def sle (x y : bitvec n) : Prop := ¬ (slt y x)
  def sge (x y : bitvec n) : Prop := sle y x

end comparison

section conversion
  variable {α : Type}

  protected def of_nat : Π (n : ℕ), nat → bitvec n
  | 0        x := nil
  | (succ n) x := of_nat n (x / 2) ++ₜ [to_bool (x % 2 = 1)]

  protected def of_int : Π (n : ℕ), int → bitvec (succ n)
  | n (int.of_nat m)          := ff :: bitvec.of_nat n m
  | n (int.neg_succ_of_nat m) := tt :: not (bitvec.of_nat n m)

  def bits_to_nat (v : list bool) : nat :=
  list.foldl (λ r b, r + r + cond b 1 0) 0 v

  protected def to_nat {n : nat} (v : bitvec n) : nat :=
  bits_to_nat (to_list v)

  protected def to_int : Π {n : nat}, bitvec n → int
  | 0        _ := 0
  | (succ n) v :=
    cond (head v)
      (int.neg_succ_of_nat $ bitvec.to_nat $ not $ tail v)
      (int.of_nat $ bitvec.to_nat $ tail v)

end conversion

private def to_string {n : nat} : bitvec n → string
| ⟨bs, p⟩ :=
  "0b" ++ (bs^.reverse^.for (λ b, if b then #"1" else #"0"))

instance (n : nat) : has_to_string (bitvec n) :=
⟨to_string⟩
end bitvec
