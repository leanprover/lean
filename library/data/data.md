data
====

Various data types.

Basic types:

* [empty](empty.lean) : the empty type
* [unit](unit.lean) : the singleton type
* [bool](bool.lean) : the boolean values
* [num](num.lean) : generic numerals
* [string](string.lean) : ascii strings
* [nat](nat/nat.md) : the natural numbers
* [fin](fin.lean) : finite ordinals
* [less_than](less_than.lean) : finite ordinals
* [int](int/int.md) : the integers
* [rat](rat/rat.md) : the rationals

Constructors:

* [prod](prod.lean) : cartesian product
* [sum](sum.lean)
* [sigma](sigma.lean) : the dependent product
* [uprod](uprod.lean) : unordered pairs
* [option](option.lean)
* [subtype](subtype.lean)
* [quotient](quotient/quotient.md)
* [squash](squash.lean) : propositional truncation
* [list](list/list.md)
* [finset](finset/finset.md) : finite sets
* [stream](stream.lean)
* [set](set/set.md)
* [vector](vector.lean)

Types with extra information:

* [fintype](fintype/fintype.md) : finite types
* [encodable](encodable.lean) : types with a coding to nat
