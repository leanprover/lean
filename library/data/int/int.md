data.int
========

The integers.

* [basic](basic.lean) : the integers, with basic operations
* [order](order.lean) : the order relations and the sign function
* [div](div.lean)     : div, mod, gcd, lcm
* [power](power.lean) 
