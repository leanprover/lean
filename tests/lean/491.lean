import data.int
open eq.ops int
variable A : Type
variables a b : A
variable H : a = b
check H⁻¹
check -1
check 1 + -2
