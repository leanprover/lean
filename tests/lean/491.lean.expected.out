H⁻¹ : b = a
-1 : ℤ
1 + -2 : ℤ
